app/vmagent/remotewrite/client.go

```diff
@@ -391,6 +392,9 @@ func (c *client) newRequest(url string, body []byte) (*http.Request, error) {
 // The function returns false only if c.stopCh is closed.
 // Otherwise it tries sending the block to remote storage indefinitely.
 func (c *client) sendBlockHTTP(block []byte) bool {
+	if len(block) == 0 {
```

Comment:
I think it would be better to skip such blocks here https://github.com/VictoriaMetrics/VictoriaMetrics/blob/56531abd562d9a4a4e8825d1a29d9e98c18ec7b8/app/vmagent/remotewrite/client.go#L297-L305
as it would prevent vmagent from spinning up a goroutine for nothing, and won't skew metrics counting the number of sent blocks.

Btw, did you discover under what circumstances block size could be 0? I wasn't able to reproduce this so far.

Reply:
Of course we can skip empty block here, but there seems no way to test it. We didn't know why the block size could be 0. This problem is solved after we restart vmagent. We can not reproduce it too. Here are full related logs
```
2024-04-26T08:19:38.277Z   info  VictoriaMetrics/app/vmagent/main.go:116 starting vmagent at ":8430"...
2024-04-26T08:19:38.280Z   info  VictoriaMetrics/lib/memory/memory.go:42 limiting caches to 640215203020 bytes, leaving 426810135348 bytes to the OS according to -memory.allowedPercent=60
2024-04-26T08:19:38.293Z   warn  VictoriaMetrics/lib/persistentqueue/persistentqueue.go:305    "vmagent-remotewrite-data/persistent-queue/1_99BFE84469E8EB07/0000000000000000" size (181981 bytes) is bigger than writer offset (181570 bytes); this may be the case on unclean shutdown (OOM, `kill -9`, hardware reset); trying to fix it by adjusting fileSize to 181570
2024-04-26T08:19:38.293Z   info  VictoriaMetrics/lib/persistentqueue/fastqueue.go:59   opened fast persistent queue at "vmagent-remotewrite-data/persistent-queue/1_99BFE84469E8EB07" with maxInmemoryBlocks=400, it contains 312 pending bytes
2024-04-26T08:19:38.293Z   info  VictoriaMetrics/app/vmagent/remotewrite/client.go:169   initialized client for -remoteWrite.url="1:secret-url"
2024-04-26T08:19:38.295Z   info  VictoriaMetrics/app/vmagent/main.go:142 started vmagent in 0.017 seconds
2024-04-26T08:19:38.295Z   info  VictoriaMetrics/lib/promscrape/scraper.go:112   reading Prometheus configs from "/etc/ovm/vmagent/scrape.yml"
2024-04-26T08:19:38.295Z   info  VictoriaMetrics/lib/httpserver/httpserver.go:99 starting http server at http://127.0.0.1:8430/
2024-04-26T08:19:38.295Z   info  VictoriaMetrics/lib/httpserver/httpserver.go:100   pprof handlers are exposed at http://127.0.0.1:8430/debug/pprof/
2024-04-26T08:19:38.310Z   info  VictoriaMetrics/lib/promscrape/config.go:120  starting service discovery routines...
2024-04-26T08:19:38.310Z   info  VictoriaMetrics/lib/promscrape/config.go:126  started service discovery routines in 0.000 seconds
2024-04-26T08:19:38.324Z   warn  VictoriaMetrics/app/vmagent/remotewrite/client.go:333   couldn't send a block with size 304 bytes to "1:secret-url": Post "http://127.0.0.1:8691": dial tcp4 127.0.0.1:8691: connect: connection refused; re-sending the block in 2.000 seconds
2024-04-26T08:19:38.324Z   info  VictoriaMetrics/lib/promscrape/scraper.go:427   static_configs: added targets: 37, removed targets: 0; total targets: 37
2024-04-26T08:19:41.341Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9900/metrics" ({instance="10.0.19.95:9900",job="octopus-exporter.endpoint"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9900/metrics": dial tcp4 10.0.19.95:9900: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:19:45.816Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:10413/metrics/services" ({instance="10.0.19.95:10413",job="svcresctld.service"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: unexpected status code returned when scraping "http://10.0.19.95:10413/metrics/services": 503; expecting 200; response body: "{\n  \"ec\": \"E_SERVICE_NOT_READY\",\n  \"data\": null,\n  \"error\": {\n   \"message\": \"service not ready\"\n  }\n}"
2024-04-26T08:19:49.156Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9914/api/v3/prometheus/netreactor" ({instance="10.0.19.95:9914",job="netreactor-exporter.netreactor"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9914/api/v3/prometheus/netreactor": dial tcp4 10.0.19.95:9914: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:20:02.071Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9916/api/v3/prometheus/l2ping_access" ({instance="10.0.19.95:9916",job="l2ping-access-exporter.l2ping_access"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9916/api/v3/prometheus/l2ping_access": dial tcp4 10.0.19.95:9916: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:20:19.157Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9914/api/v3/prometheus/netreactor" ({instance="10.0.19.95:9914",job="netreactor-exporter.netreactor"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9914/api/v3/prometheus/netreactor": dial tcp4 10.0.19.95:9914: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:20:32.071Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9916/api/v3/prometheus/l2ping_access" ({instance="10.0.19.95:9916",job="l2ping-access-exporter.l2ping_access"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9916/api/v3/prometheus/l2ping_access": dial tcp4 10.0.19.95:9916: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:20:49.157Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9914/api/v3/prometheus/netreactor" ({instance="10.0.19.95:9914",job="netreactor-exporter.netreactor"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9914/api/v3/prometheus/netreactor": dial tcp4 10.0.19.95:9914: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:21:02.072Z   warn  VictoriaMetrics/lib/promscrape/scrapework.go:376   cannot scrape target "http://10.0.19.95:9916/api/v3/prometheus/l2ping_access" ({instance="10.0.19.95:9916",job="l2ping-access-exporter.l2ping_access"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: error when scraping "http://10.0.19.95:9916/api/v3/prometheus/l2ping_access": dial tcp4 10.0.19.95:9916: connect: connection refused; try -enableTCP6 command-line flag if you scrape ipv6 addresses
2024-04-26T08:21:11.302Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #1: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 2.000 seconds
2024-04-26T08:21:11.302Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #1: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 2.000 seconds
2024-04-26T08:21:11.303Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #1: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 2.000 seconds
2024-04-26T08:21:11.303Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #1: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 2.000 seconds
2024-04-26T08:21:13.304Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #2: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 4.000 seconds
2024-04-26T08:21:13.304Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #2: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 4.000 seconds
2024-04-26T08:21:13.304Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #2: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 4.000 seconds
2024-04-26T08:21:13.304Z   error   VictoriaMetrics/app/vmagent/remotewrite/client.go:383   unexpected status code received after sending a block with size 0 bytes to "1:secret-url" during retry #2: 422; response body="{\"code\":422,\"message\":\"Failed decompressing payload with snappy decoder.\"}"; re-sending the block in 4.000 seconds
```

Reply:
The test could be done easily if we introduce function/method level mocking via a 3rd-party package. 
[gomonkey](https://github.com/agiledragon/gomonkey) would be a great choice, tiny, with many active users.

I would like to recommend it again, while also ensuring that this package (hopefully) complies with the contribution guidelines.
- Avoid [big external dependencies](https://medium.com/@valyala/stripping-dependency-bloat-in-victoriametrics-docker-image-983fb5912b0d).

Reply:
With [gomonkey](https://github.com/agiledragon/gomonkey), I can patch `MustReadBlock` to test it. For example:
```go
func TestClientSendEmptyBlockHTTP(t *testing.T) {
	srv := newTestRemoteWriteServer()
	path := "test"
	defer mustDeleteDir(path)
	fq := persistentqueue.MustOpenFastQueue(path, "test", 100, 0, false)
	defer fq.MustClose()
	var p = gomonkey.ApplyMethod(reflect.TypeOf(fq), "MustReadBlock", func(_ *persistentqueue.FastQueue, dst []byte) ([]byte, bool) {
		fmt.Println("returning empty block")
		return nil, true
	})
	defer p.Reset()
	url := srv.URL + "/api/v1/write"
	c := newHTTPClient(1, url, url, fq, 1)
	argIdx := 1
	limitReached := metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_rate_limit_reached_total{url=%q}`, c.sanitizedURL))
	if bytesPerSec := rateLimit.GetOptionalArg(argIdx); bytesPerSec > 0 {
		logger.Infof("applying %d bytes per second rate limit for -remoteWrite.url=%q", bytesPerSec, url)
		c.rl = ratelimiter.New(int64(bytesPerSec), limitReached, c.stopCh)
	}
	c.bytesSent = metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_bytes_sent_total{url=%q}`, c.sanitizedURL))
	c.blocksSent = metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_blocks_sent_total{url=%q}`, c.sanitizedURL))
	c.rateLimit = metrics.GetOrCreateGauge(fmt.Sprintf(`vmagent_remotewrite_rate_limit{url=%q}`, c.sanitizedURL), func() float64 {
		return float64(rateLimit.GetOptionalArg(argIdx))
	})
	c.requestDuration = metrics.GetOrCreateHistogram(fmt.Sprintf(`vmagent_remotewrite_duration_seconds{url=%q}`, c.sanitizedURL))
	c.requestsOKCount = metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_requests_total{url=%q, status_code="2XX"}`, c.sanitizedURL))
	c.errorsCount = metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_errors_total{url=%q}`, c.sanitizedURL))
	c.packetsDropped = metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_packets_dropped_total{url=%q}`, c.sanitizedURL))
	c.retriesCount = metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_remotewrite_retries_count_total{url=%q}`, c.sanitizedURL))
	c.sendDuration = metrics.GetOrCreateFloatCounter(fmt.Sprintf(`vmagent_remotewrite_send_duration_seconds_total{url=%q}`, c.sanitizedURL))
	metrics.GetOrCreateGauge(fmt.Sprintf(`vmagent_remotewrite_queues{url=%q}`, c.sanitizedURL), func() float64 {
		return float64(*queues)
	})
	go func() {
		c.runWorker()
	}()
	if badRequestsCount != 1 {
		t.Fatalf("got %d bad requests, want 1", badRequestsCount)
	}
	if normalRequestsCount != 0 {
		t.Fatalf("got %d normal requests, want 0", normalRequestsCount)
	}
}
```